README.md
# Riak Test Server

Basho has (for good reasons) discontinued support for test servers in their
client libraries, but it turns out the critical functionality is built into Riak
regardless. This project wraps some drop in code (currently just Ruby) around a
Docker container that gives the same functionality and allows rapidly resetting
the Riak backend inbetween tests.

## Pre-requisites

* [docker](https://docs.docker.com/installation/)
* A working docker host environment (I like [dvm](https://github.com/fnichol/dvm) on OS X)
* A hostname of `docker` mapped to the IP of your container host (192.168.42.43 for dvm)
* At least 1Gb available to the docker VM. The base Erlang runtime seems to be a bit of a memory hog. (`export DOCKER_MEMORY=1024` for dvm)

## Installation

Add this line to your application's Gemfile and `bundle`:

    gem 'riak_test_server'

Or install it yourself as:

    $ gem install riak_test_server

## Usage

### Setup

Near the top of `test_helper.rb` or the like:

```
require "riak_test_server"
RiakTestServer.setup
```

### Clearing

Wherever you want to clear out Riak, say in `setup`:

```
RiakTestServer.clear
```

### Options

You can override a bunch of defaults by passing options to `RiakTestServer.setup`:

* `:container_host` - hostname or IP of the container host [default: `"docker"`]
* `:http_port` - port to expose Riak HTTP on on the `container_host` [default: `"8098"`]
* `:pb_port` - port to expose Riak Protocol Buffers on on the `container_host` [default: `"8087"`]
* `:container_name` - name to use for the container when `docker run`-ing it; handy to change if you want different containers for different apps [default: `"riak_test_server"`]
* `:repository` - supplier of docker images [default: `"ntalbott/riak_test_server"`]
* `:tag` - docker image to use; `docker search ntalbott/riak_test_server` to see if other Riak versions are available [default: `"latest"`]
* `:docker_bin` - name of the docker binary [default: `"docker"`]
* `:force_restart` - will force the hard restart of any running container (slow) [efault: `false`]


CHANGELOG.md
# 1.1.2

* Increase container communication timeout.

# 1.1.1

* Kill stopped containers even when not :force_restart.

# 1.1.0

* Better timeouts for docker commands.
* Only restart a running container if `:force_restart` is passed.

# 1.0.0

* Initial release


test/riak_test_server_tests.rb
require "excon"

require "riak_test_server"

module RiakTestServerTests
  def setup
    RiakTestServer.setup(server_options)
  end

  def test_clear
    response = Excon.put(put_url("1"), body: "one", headers: {"Content-Type" => "text/plain"})
    assert_equal 204, response.status, response.body
    assert_equal "one", Excon.get(get_url("1")).body

    RiakTestServer.clear
    assert_equal "not found", Excon.get(get_url("1")).body.chomp
  end

  def put_url(key)
    "http://docker:8098/buckets/a/keys/#{key}"
  end

  def get_url(key)
    "http://docker:8098/buckets/a/keys/#{key}?r=1"
  end

  def server_options
    {
      container_name: "riak_test_server_tests",
      docker_tag: version
    }
  end
end


test/riak_test_server_12_test.rb
require "minitest/autorun"
require "riak_test_server_tests"

class RiakTestServer12Test < Minitest::Test
  include RiakTestServerTests

  def version
    "1.2"
  end
end


test/riak_test_server_13_test.rb
require "minitest/autorun"
require "riak_test_server_tests"

class RiakTestServer13Test < Minitest::Test
  include RiakTestServerTests

  def version
    "1.3"
  end
end


ruby/riak_test_server.rb
require "expect"
require "timeout"
require "open3"
require "json"

module RiakTestServer
  class Error < StandardError; end

  def self.setup(options={})
    @server.stop if @server
    @server = Server.new(options)

    @server.setup
    @server.start
    @server.check
    @server.clear
  end

  def self.clear
    @server.clear
  end

  class Server
    attr_reader :repository, :tag
    def initialize(options)
      @force_restart = (options[:force_restart] || false)
      @container_host = (options[:container_host] || "docker")
      @container_name = (options[:container_name] || "riak_test_server")
      @http_port = (options[:http_port] || "8098")
      @pb_port = (options[:pb_port] || "8087")
      @repository = (options[:docker_repository] || "ntalbott/riak_test_server")
      @tag = (options[:docker_tag] || "latest")
      @docker_bin = (options[:docker_bin] || "docker")
    end

    def start
      details = JSON.parse(docker("inspect #{@container_name}", allowed_exits: [0, 1]))
      if details.size > 0
        detail = details.first
        running = (detail["State"] && detail["State"]["Running"])
        if(@force_restart || !running)
          stop
          docker "rm #{@container_name}"
        else
          return
        end
      end

      docker %W(
        run
        --name #{@container_name}
        --detach=true
        --interactive=true
        --publish=#{@http_port}:8098
        --publish=#{@pb_port}:8087
        #{@repository}:#{@tag}
      ).join(" "), timeout: 5
    end

    def stop
      @console_io.close if @console_io
      docker "stop #{@container_name}", timeout: 15
    end

    def setup
      unless docker("images", timeout: 5) =~ /#{repository}\s+#{tag}/
        docker "pull #{repository}:#{tag}", timeout: 60
      end
    end

    def clear
      result = console("{riak_kv_memory_backend:reset(), boom}.").chomp.strip
      raise "Unable to reset backend (#{result})" unless result == "{ok,boom}"
    end

    def check
      retries = 20
      loop do
        result = console("riak_kv_console:vnode_status([]).")
        break if result.split(/^VNode: \w+$/i).size >= 4 # at least 3 vnodes are available

        raise "vnodes not starting in time" if retries == 0
        retries -= 1
        sleep 1
      end
    end

    private

    def docker(command, options={})
      full_command = "#{@docker_bin} #{command}"
      timeout = (options[:timeout] || 1)
      allowed_exits = (options[:allowed_exits] || [0])
      Timeout.timeout(timeout) do
        stdout, stderr, status = Open3.capture3(full_command)
        unless(allowed_exits.include?(status.exitstatus))
          raise [
            "#{full_command} exited with unexpected status #{status.exitstatus} (expected: #{allowed_exits.inspect})",
            "STDOUT: #{stdout}",
            "STDERR: #{stderr}"
          ].join("\n")
        end
        stdout
      end
    rescue Timeout::Error
      raise RiakTestServer::Error, "Timed out running `#{full_command}` after #{timeout} seconds; is your Docker host running?"
    end

    PROMPT = /\(riak@[\w\.]+\)(\d+)>\s*/
    def console(command)
      raise "Command not terminated with a `.`: #{command}" unless command =~ /\.\s*\z/
      attach do |io|
        io.puts(command)
        response = io.expect(PROMPT, 60)
        if response
          PROMPT.match(response.first).pre_match
        else
          raise "Prompt not returned after sending #{command} to Riak console"
        end
      end
    end

    def attach
      unless @console_io
        @console_io = IO.popen([@docker_bin, "attach", @container_name], "r+", err: :out).tap{|io| io.sync = true}
        @console_io.puts("ok.")
        unless @console_io.expect(PROMPT, 60)
          raise "Unable to get a prompt on the console: @console_io.read"
        end
      end

      yield(@console_io)
    end
  end
end
